src/Pirouette/Term/Syntax/Base.hs

```diff
@@ -31,60 +23,65 @@ import           Data.Typeable
 import qualified Data.Text as T
 import qualified Data.Map as M
 import           Data.Data
+import Data.Void
+
+type EqOrdShowDataTypeable a = (Eq a, Ord a, Show a, Data a, Typeable a)
+
+type LanguageConstrs lang
+  = (EqOrdShowDataTypeable (BuiltinTypes lang)
+    , EqOrdShowDataTypeable (BuiltinTerms lang)
+    , EqOrdShowDataTypeable (Constants lang))
 
+type PrettyLang lang = (Pretty (BuiltinTypes lang), Pretty (BuiltinTerms lang), Pretty (Constants lang))
 
--- * PIR Types
-
--- |Builtin Plutus Types
-data PIRType
-  = PIRTypeInteger
-  | PIRTypeByteString
-  | PIRTypeUnit
-  | PIRTypeBool
-  | PIRTypeString
-  | PIRTypeData
-  | PIRTypeList (Maybe PIRType)
-  | PIRTypePair (Maybe PIRType) (Maybe PIRType)
-  deriving (Eq, Ord, Show, Data, Typeable)
-
-defUniToType :: forall k (a :: k) . DefaultUni (P.Esc a) -> PIRType
-defUniToType DefaultUniInteger     = PIRTypeInteger
-defUniToType DefaultUniByteString  = PIRTypeByteString
-defUniToType DefaultUniUnit        = PIRTypeUnit
-defUniToType DefaultUniBool        = PIRTypeBool
-defUniToType DefaultUniString      = PIRTypeString
-defUniToType DefaultUniData        = PIRTypeData
-defUniToType (DefaultUniList a)    = PIRTypeList (Just (defUniToType a))
-defUniToType (DefaultUniPair a b)  = PIRTypePair (Just $ defUniToType a) (Just $ defUniToType b)
-defUniToType DefaultUniProtoList   = PIRTypeList Nothing
-defUniToType DefaultUniProtoPair   = PIRTypePair Nothing Nothing
-defUniToType (DefaultUniApply DefaultUniProtoPair a) = PIRTypePair (Just $ defUniToType a) Nothing
+class (Data lang, Typeable lang, LanguageConstrs lang) => LanguageDef lang where
+  type BuiltinTypes lang :: *
+  type BuiltinTerms lang :: *
+  type Constants lang :: *
```

Comment:
Interesting question. I'm not entirely sure that they should. That would mean we can't reuse certain types for different languages. Type inference has not been a problem yet so I'd advocate not touching it. That being said, its a great place for a comment with respect to that. It will freshen our memory if and when we get inference problems.